src/stores/TagOrderStore.js

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2017 Vector Creations Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import {Store} from 'flux/utils';
+import dis from '../dispatcher';
+import Analytics from '../Analytics';
+import MatrixClientPeg from "../MatrixClientPeg";
+
+const INITIAL_STATE = {
+    orderedTags: null,
+    allTags: null,
+};
+
+/**
+ * A class for storing application state for ordering tags in the TagPanel.
+ */
+class TagOrderStore extends Store {
+    constructor() {
+        super(dis);
+
+        // Initialise state
+        this._state = INITIAL_STATE;
+    }
+
+    _setState(newState) {
+        this._state = Object.assign(this._state, newState);
+        this.__emitChange();
+    }
+
+    __onDispatch(payload) {
+        switch (payload.action) {
+            // Get ordering from account data, once the client has synced
+            case 'sync_state':
+                if (payload.prevState === "PREPARED" && payload.state === "SYNCING") {
+                    const accountDataEvent = MatrixClientPeg.get().getAccountData('im.vector.web.tag_ordering');
+
+                    const orderedTags = accountDataEvent && accountDataEvent.getContent() ?
+                        accountDataEvent.getContent().tags : null;
+
+                    this._setState({orderedTags});
+                }
+            break;
+            // Initialise the state such that if account data is unset, default to the existing ordering
+            case 'all_tags':
+                this._setState({
+                    allTags: payload.tags.sort(), // Sort lexically
```

Comment:
I don't think this will do the right thing when you join a new group: the new group will be added to `allTags` but won't propagate to `orderedTags` which is what the TagPanel reads, so they new group won't show up? Somewhere we'll have to work out what changed in the new set of tags and make the equivalent change to the `orderedTags`, preserving the order.

I assume the idea here is that the TagOrderStore knows nothing about groups, but it doesn't seem great that it's the responsibility of a component to give the store the set of all tags, ie. that component has to be mounted once (and only once?) in the view for the store to work. I'd guess the pattern I'd expect here would be to have another store or something for the complete set of tags, then the TagOrderStore consumes that and exposes the same info but ordered.

Reply:
>it's the responsibility of a component to give the store the set of all tags

I could move the fetch to get the list of groups to the TagOrderStore. The issue is that this would be perfect in an action creator, something that we haven't done before. Perhaps we should go and make a bunch of action creators and formalises all of our existing dispatches by wrapping them in functions (which is standard in other apps AFAICT).

>have another store or something for the complete set of tags, then the TagOrderStore consumes that

Stores really should not communicate. `all_tags` is already being consumed by another store, and I believe the general pattern is to duplicate state as required such that the state of the store really only depends on its own previous state and the actions it receives.

Reply:
>I don't think this will do the right thing when you join a new group

This is true. If only our store could listen for the group membership update...

Reply:
Having TagOrderStore fetch the list of groups is fine by me